File: 백준/Silver/15664. N과 M （10）/N과 M （10）.js
const [[N, M], nums] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
const path = new Array(M);
const ans = [];
nums.sort((a, b) => a - b);
DFS(0, 0);
console.log(ans.map(line => line.join(' ')).join('\n'));
function DFS(start, depth) {
  if(depth === M) {
    ans.push([...path]);
    return;
  }
  let before = Number.MIN_VALUE;
  for(let i = start; i < N; i++) {
    if(before === nums[i]) continue;
    path[depth] = nums[i];
    before = nums[i];
    DFS(i + 1, depth + 1);
  }
}

File: 백준/Silver/15664. N과 M （10）/N과 M （10）.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[] nums, path;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        nums = new int[N];
        path = new int[M];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(nums);
        DFS(0, 0);
        System.out.print(sb);
    }
    static void DFS(int start, int depth) {
        if(depth == M) {
            for(int i = 0; i < M; i++) {
                sb.append(path[i]);
                if(i != M - 1) sb.append(" ");
            }
            sb.append("\n");
            return;
        }
        int before = Integer.MIN_VALUE;
        for(int i = start; i < N; i++) {
            if(before == nums[i]) continue;
            path[depth] = nums[i];
            before = nums[i];
            DFS(i + 1, depth + 1);
        }
    }
}
